File: backend/internal/models/report_export.go
package models

import "time"

// AttendanceReportRequest drives the attendance PDF export — one row per
// attendance record for the class across the date range.
type AttendanceReportRequest struct {
	ClassID string    `form:"class_id" binding:"required"`
	From    time.Time `form:"from" binding:"required" time_format:"2006-01-02"`
	To      time.Time `form:"to" binding:"required" time_format:"2006-01-02"`
	// Columns is an optional subset of: date, student, index_number, status, note.
	// Empty means "all columns".
	Columns []string `form:"columns"`
}

// MarksReportRequest drives the marks PDF export — one row per student's
// mark for a class+term+subject.
type MarksReportRequest struct {
	ClassID   string `form:"class_id" binding:"required"`
	TermID    string `form:"term_id" binding:"required"`
	SubjectID string `form:"subject_id" binding:"required"`
	// Columns is an optional subset of: student, index_number, marks, max_marks, percentage.
	Columns []string `form:"columns"`
}

File: frontend/src/services/reportExport.ts
import api from "../lib/api";

export const ATTENDANCE_REPORT_COLUMNS = [
  { value: "date", label: "Date" },
  { value: "student", label: "Student" },
  { value: "index_number", label: "Index No." },
  { value: "status", label: "Status" },
  { value: "note", label: "Note" },
];

export const MARKS_REPORT_COLUMNS = [
  { value: "student", label: "Student" },
  { value: "index_number", label: "Index No." },
  { value: "marks", label: "Marks" },
  { value: "max_marks", label: "Max Marks" },
  { value: "percentage", label: "%" },
];

async function downloadPdf(url: string, params: Record<string, unknown>, filename: string) {
  const response = await api.get<Blob>(url, { params, responseType: "blob" });
  const blobUrl = window.URL.createObjectURL(response.data);
  const link = document.createElement("a");
  link.href = blobUrl;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  link.remove();
  window.URL.revokeObjectURL(blobUrl);
}

export const reportExportApi = {
  exportAttendance: (params: { class_id: string; from: string; to: string; columns?: string[] }) =>
    downloadPdf("/reports/attendance", params, "attendance-report.pdf"),

  exportMarks: (params: { class_id: string; term_id: string; subject_id: string; columns?: string[] }) =>
    downloadPdf("/reports/marks", params, "marks-report.pdf"),
};
